src/core/public/overlays/banners/banners_service.tsx

```diff
@@ -0,0 +1,141 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { sortBy } from 'lodash';
+import React from 'react';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+import { BannersList } from './banners_list';
+
+/**
+ * A function that will unmount the banner from the element.
+ * @public
+ */
+export type OverlayBannerUnmount = () => void;
+
+/**
+ * A function that will mount the banner inside the provided element.
+ * @param element an element to render into
+ * @returns a {@link OverlayBannerUnmount}
+ * @public
+ */
+export type OverlayBannerMount = (element: HTMLElement) => OverlayBannerUnmount;
+
+/** @public */
+export interface OverlayBannersStart {
+  /**
+   * Add a new banner
+   *
+   * @param mount {@link OverlayBannerMount}
+   * @param priority optional priority order to display this banner. Higher priority values are shown first.
+   * @returns a unique symbol for the given banner to be used with {@link OverlayBannersStart.remove} and
+   *          {@link OverlayBannersStart.replace}
+   */
+  add(mount: OverlayBannerMount, priority?: number): symbol;
+
+  /**
+   * Remove a banner
+   *
+   * @param id the unique symbol for the banner returned by {@link OverlayBannersStart.add}
+   * @returns if the banner was found or not
+   */
+  remove(id: symbol): boolean;
+
+  /**
+   * Replace a banner in place
+   *
+   * @param id the unique symbol for the banner returned by {@link OverlayBannersStart.add}
+   * @param mount {@link OverlayBannerMount}
+   * @param priority optional priority order to display this banner. Higher priority values are shown first.
+   * @returns a new symbol for the given banner to be used with {@link OverlayBannersStart.remove} and
+   *          {@link OverlayBannersStart.replace}
+   */
+  replace(id: symbol, mount: OverlayBannerMount, priority?: number): symbol;
+
+  /** @internal */
+  get$(): Observable<OverlayBanner[]>;
+  getComponent(): JSX.Element;
+}
+
+/** @internal */
+export interface OverlayBanner {
+  id: symbol;
+  mount: OverlayBannerMount;
+  priority: number;
+}
+
+/** @internal */
+export class OverlayBannersService {
+  public start(): OverlayBannersStart {
+    const banners$ = new BehaviorSubject<ReadonlyMap<symbol, OverlayBanner>>(new Map());
+
+    const addToMap = (banners: ReadonlyMap<symbol, OverlayBanner>, bannerToAdd: OverlayBanner) => {
+      return new Map(
```

Comment:
It looks like we can extract `Priority queue` out from the banner service. That reduces the half of logic in the file and make testing focused on service logic.